### PR TITLE
Custom sheet testing

### LINKED ISSUE #82: Custom Sheet Testing
Description: We want to incorporate custom sheets which can be scanned by Scantron AI Algorithms.

Given our Ai API
when custom sheets are submitted 
then it should result in efficiently extracted csv file.

Developer Notes: 
You have to modify BubbleScan Ai directory.

Success Criteria:
 
- [x] #97
- [x] #104
- [x] Returns CSV files.
- [x] At least 98% Accuracy for the custom sheets wrt Ground Truth.

### PR DESCRIPTION
**Pull request by ramezmosad**
Fixes #82 

**What was changed?**

In this update, the backend Flask server (AppServer.py) was modified to handle custom scantron sheets using the CustomProcessor class. This change primarily affects the file upload, processing, and CSV generation routes in AppServer.py. Specifically:

* I added logic to dynamically select CustomProcessor or Scantron95945 based on the type of sheet uploaded.
* Thanks to @kate-holdener, we ensured that the correct paths were set up for BubbleScan_AI so that the application could locate CustomProcessor.

**Why was it changed?**

This change was implemented to address an issue (#82) where the application did not support processing custom scantron sheets. Users needed the ability to upload both standard and custom sheets, with the system dynamically adapting to the type of sheet uploaded. Without this support, attempting to process custom sheets resulted in errors, such as "module not found" or "CustomProcessor not defined." By making these changes, the application can process custom sheets and generate corresponding CSV files as expected.

**How was it changed?**

1. sys.path.append was adjusted to correctly locate the BubbleScan_AI module, by moving it to the AppServer directory. This ensures that Python can find and import CustomProcessor and Scantron95945.

2. In the process_pdf function in AppServer.py, we added logic to select either CustomProcessor or Scantron95945 based on the sheetType parameter. This enables the backend to handle both standard and custom sheets dynamically.
python

3. After selecting the appropriate processor, the application now extracts responses and generates CSV data. This CSV data is saved and made available for download, addressing the "CSV file not found" error encountered previously.

### CODE REVIEW

**Comment by Chirag2x** [comment]
@ramezmosad, Well the application shows it's working. But when downloading the file I am getting this error. Can you look into it. That would be much appreciated!
Screenshot 2024-11-11 at 5 08 07 PM

File: ServerCode/application/AppServer.py [hunk 1]
```diff
@@ -161,31 +155,40 @@ def file_upload(self):
         else:
             return jsonify({"status": "error", "message": "Only PDF files are allowed"})
 
-    def process_pdf(self, pdf_file, file_id):
+
+    def process_pdf(self, pdf_file, file_id, sheet_type):
```
Comment (mylescorrea) [inline ServerCode/application/AppServer.py:159]: nice job implementing how the pdf will be processed depending on the custom or regular scantron

Comment (mylescorrea) [inline ServerCode/application/AppServer.py:165]: nice job implementing how the pdf will be processed depending on the custom or regular scantron

**Review by mylescorrea — COMMENTED** [review]
Right now I know it's not functional yet, but I see the framework for the custom exam sheet processing is going well. I can't wait to see the results of the AI component when it is finished!

**Comment by yildizahmetali** [comment]
Great Work Ramez! You have a nice framework, the program can differ if it has custom sheet processing or scantron processing. I believe you will do a great job of finishing the full implementation.

**Comment by kate-holdener** [comment]
> @ramezmosad, Well the application shows it's working. But when downloading the file I am getting this error. Can you look into it. That would be much appreciated! Screenshot 2024-11-11 at 5 08 07 PM

I get the same error. It looks like it's really close to working!

**Comment by Chirag2x** [comment]
@ramezmosad, The custom sheet is processing extremely well. You pulled this off. But here is a thing, Now the scantron sheet is not working as it should. I am getting blank CSV file as an output.
Screenshot 2024-12-01 at 1 34 55 PM

File: ServerCode/application/AppServer.py [hunk 2]
```diff
@@ -2,38 +2,21 @@
...
 import webbrowser
-import sys
-
-# Third-party imports
-from flask import Flask, jsonify, redirect, request, send_from_directory, url_for
-from flask_cors import CORS
 from werkzeug.utils import secure_filename
+from flask import Flask, request, jsonify, send_from_directory, redirect, url_for
+from flask_cors import CORS
+from BubbleScan_AI.Scantron import Scantron95945
```
Comment (kate-holdener) [inline ServerCode/application/AppServer.py:13]: @ramezmosad I think the problem @Chirag2x reported has to do with this import. Try using this instead:
```suggestion
from application.Scantron import Scantron95945
```
As you mentioned, there are two copies of Scantron.py, and I think we accidentally switched to using the wrong copy.

Comment (Chirag2x) [reply ServerCode/application/AppServer.py:13]: I tried this solution earlier but it's still not working. The issue might be something else.

**Review by Chirag2x — APPROVED** [review]
Good Job @ramezmosad. You have done an amazing efforts to complete this, and it works the way it should.

**Pull request merged**